File: OC/DCUtils/DCUIKit.m
//
//  DCUIKit.m
//  FYJ
//
//  Created by ma on 2018/7/21.
//  Copyright © 2018年 issuser. All rights reserved.
//

#import "DCUIKit.h"
#import "LoginViewController.h"
#import "AppNavigationController.h"
@implementation DCUIKit
+ (CAShapeLayer *)dc_maskWithRect:(CGRect)rect {
    UIBezierPath *maskPath = [UIBezierPath bezierPathWithRoundedRect:rect      byRoundingCorners:UIRectCornerAllCorners   cornerRadii:CGSizeMake(rect.size.width/2.0,rect.size.width/2.0)];
    CAShapeLayer *maskLayer = [[CAShapeLayer alloc] init];
    maskLayer.frame = rect;
    maskLayer.path = maskPath.CGPath;
    return maskLayer;
}
+ (CAShapeLayer *)dc_maskWithRect:(CGRect)rect radiiSize:(CGSize)radiiSize{
    UIBezierPath *maskPath = [UIBezierPath bezierPathWithRoundedRect:rect      byRoundingCorners:UIRectCornerAllCorners   cornerRadii:radiiSize];
    CAShapeLayer *maskLayer = [[CAShapeLayer alloc] init];
    maskLayer.frame = rect;
    maskLayer.path = maskPath.CGPath;
    return maskLayer;
}
+ (CAShapeLayer *)dc_maskWithRect:(CGRect)rect rectCorners:(UIRectCorner)rectCorners radiiSize:(CGSize)radiiSize {
    UIBezierPath *maskPath = [UIBezierPath bezierPathWithRoundedRect:rect      byRoundingCorners:rectCorners cornerRadii:radiiSize];
    CAShapeLayer *maskLayer = [[CAShapeLayer alloc] init];
    maskLayer.frame = rect;
    maskLayer.path = maskPath.CGPath;
    return maskLayer;
}
+ (CAShapeLayer *)dc_maskWithRect:(CGRect)rect borderColor:(UIColor *)borderColor borderWidth:(CGFloat)borderWidth {
     UIBezierPath *maskPath = [UIBezierPath bezierPathWithRoundedRect:rect      byRoundingCorners:UIRectCornerAllCorners   cornerRadii:CGSizeMake(5,5)];
    CAShapeLayer *borderLayer = [CAShapeLayer layer];
    borderLayer.path    =   maskPath.CGPath;
    borderLayer.fillColor  = [UIColor clearColor].CGColor;
    borderLayer.strokeColor    = borderColor.CGColor;
    borderLayer.lineWidth      = borderWidth;
    return borderLayer;
}
+ (CGFloat)compressImage:(UIImage *)sourceImage {
    CGFloat f = UIImageJPEGRepresentation(sourceImage, 1).length/1024.0;
    NSLog(@">>>>>>>%.2f",(CGFloat)(f/1024));
    float factor = 1.0f;
    NSData *data = UIImageJPEGRepresentation(sourceImage, factor);
    while (data.length/1024/1024  >= 1.00f) { //图片超过1M 压缩
        factor -= 0.1f;
        if (factor <= 0.0f) {
            break;
        }
        data = UIImageJPEGRepresentation(sourceImage, factor);
    }
    return factor;
}
+ (UIImage *)dc_makeImageWithView:(UIView *)view withSize:(CGSize)size {
    UIGraphicsBeginImageContextWithOptions(size, NO,0.0);
    [view.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}
+ (UIImage *)dc_scaleToSize:(UIImage *)img size:(CGSize)size {
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0);
    [img drawInRect:CGRectMake(0, 0, size.width, size.height)];
    UIImage* scaledImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return scaledImage;
}
+ (UIImage *)dc_synthesisImageWithImage:(UIImage *)firstImage otherImage:(UIImage *)otherImage {
    CGFloat newImageW = firstImage.size.width + otherImage.size.width;
    CGFloat newImageH = firstImage.size.height > otherImage.size.height?firstImage.size.height:otherImage.size.height;
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(newImageW, newImageH), NO, 0.0);
    [firstImage drawInRect:CGRectMake(0, 0, firstImage.size.width, firstImage.size.height)];
    [otherImage drawInRect:CGRectMake(firstImage.size.width - 2, firstImage.size.height - otherImage.size.height, otherImage.size.width, otherImage.size.height)];
    UIImage *contactImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return contactImage;
}
+ (UIImage *)dc_synthesisImageWithImage:(UIImage *)topImage bottomImage:(UIImage *)bottomImage {
    CGFloat newImageW = topImage.size.width > bottomImage.size.width?topImage.size.width:bottomImage.size.width;
    CGFloat newImageH = topImage.size.height + bottomImage.size.height;
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(newImageW, newImageH), NO, 0.0);
    [topImage drawInRect:CGRectMake(0, 0, topImage.size.width, topImage.size.height)];
    [bottomImage drawInRect:CGRectMake((newImageW-bottomImage.size.width)/2.0, topImage.size.height, bottomImage.size.width, bottomImage.size.height)];
    UIImage *contactImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return contactImage;
}
+ (void)gotoLogin {
    UIWindow *keyWindow  = [UIApplication sharedApplication].keyWindow;
    UIViewController *controller = keyWindow.rootViewController;
    LoginViewController *vc = [[LoginViewController alloc]init];
    vc.hidesBottomBarWhenPushed = YES;
    AppNavigationController *logingNav = [[AppNavigationController alloc]initWithRootViewController:vc];
    [controller presentViewController:logingNav animated:YES completion:nil];
}
@end

File: Swift/DCUtils/DCUIKit.swift
//
//  DCUIKit.swift
//  SwiftDemo
//
//  Created by ma on 2018/11/12.
//  Copyright © 2018年 ma. All rights reserved.
//

import UIKit
class DCUIKit: NSObject {
    
    /// 设置圆形遮罩
    ///
    /// - Parameter rect: 目标bounds
    /// - Returns: CAShapeLayer对象
    class func dc_maskWithRect(rect:CGRect) -> CAShapeLayer {
        let maskPath = UIBezierPath.init(roundedRect: rect, cornerRadius: rect.size.width/2.0)
        let maskLayer = CAShapeLayer.init()
        maskLayer.frame = rect
        maskLayer.path = maskPath.cgPath
        return maskLayer
    }
    /// 设置圆形遮罩
    ///
    /// - Parameter radiiSize: 圆角大小
    /// - Returns: CAShapeLayer对象
    class func dc_maskWithRect(rect:CGRect,radiiSize:CGSize) -> CAShapeLayer {
        let maskPath = UIBezierPath.init(roundedRect: rect, byRoundingCorners: UIRectCorner.allCorners, cornerRadii: CGSize.init(width: radiiSize.width, height: radiiSize.height))
        let maskLayer = CAShapeLayer.init()
        maskLayer.frame = rect
        maskLayer.path = maskPath.cgPath
        return maskLayer
    }
    /// 设置圆形遮罩
    /// - Parameter rectCorners: 圆角范围
    /// - Parameter radiiSize: 圆角大小
    /// - Returns: CAShapeLayer对象
    class func dc_maskWithRect(rect:CGRect,rectCorners:UIRectCorner,radiiSize:CGSize) -> CAShapeLayer {
        let maskPath = UIBezierPath.init(roundedRect: rect, byRoundingCorners: rectCorners, cornerRadii: CGSize.init(width: radiiSize.width, height: radiiSize.height))
        let maskLayer = CAShapeLayer.init()
        maskLayer.frame = rect
        maskLayer.path = maskPath.cgPath
        return maskLayer
    }
    
    /// 设置圆角外环
    ///
    /// - Parameters:
    ///   - borderColor: 外环颜色
    ///   - borderWidth: 外环宽度
    /// - Returns: CAShapeLayer对象
    class func dc_maskWithRect(rect:CGRect,borderColor:UIColor,borderWidth:CGFloat) -> CAShapeLayer {
        let maskPath = UIBezierPath.init(roundedRect: rect, byRoundingCorners: UIRectCorner.allCorners, cornerRadii: CGSize.init(width: 5, height: 5))
        let borderLayer = CAShapeLayer.init()
        borderLayer.path = maskPath.cgPath
        borderLayer.fillColor  = UIColor.clear.cgColor
        borderLayer.strokeColor = borderColor.cgColor
        borderLayer.lineWidth = borderWidth;
        return borderLayer
    }
    /// 根据View生成图片
    class func dc_makeImage(view:UIView,size:CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(size, false, 0)
        view.layer.render(in: UIGraphicsGetCurrentContext()!)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
    /// 根据Size缩放图片
    class func dc_scaleToSize(image:UIImage?,size:CGSize) -> UIImage? {
        if image == nil {
            assertionFailure("图片为空")
            return nil
        }
        UIGraphicsBeginImageContextWithOptions(size, false, 0)
        image!.draw(in: CGRect.init(x: 0, y: 0, width: size.width, height: size.height))
        let scaledImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return scaledImage
    }
    /// 两张图片合成一张图片,左右合成
    class func dc_synthesisImage(firstImage:UIImage?,otherImage:UIImage?) -> UIImage? {
        if firstImage == nil || otherImage == nil {
            assertionFailure("图片为空")
            return nil
        }
        let newImageW = firstImage!.size.width + otherImage!.size.width;
        let newImageH = firstImage!.size.height > otherImage!.size.height ?firstImage!.size.height:otherImage!.size.height
         UIGraphicsBeginImageContextWithOptions(CGSize.init(width: newImageW, height: newImageH), false, 0)
        firstImage!.draw(in: CGRect.init(x: 0, y: 0, width: firstImage!.size.width, height: firstImage!.size.height))
        otherImage!.draw(in: CGRect.init(x: firstImage!.size.width - 2, y: firstImage!.size.height - otherImage!.size.height, width: otherImage!.size.width, height: otherImage!.size.height))
        let contactImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext();
        return contactImage;
    }
    /// 两张图片合成一张图片,上下合成
    class func dc_synthesisImage(topImage:UIImage?,bottomImage:UIImage?) -> UIImage? {
        if topImage == nil || bottomImage == nil {
            assertionFailure("图片为空")
            return nil
        }
        let newImageW =  topImage!.size.width > bottomImage!.size.width ?topImage!.size.width:bottomImage!.size.width
        let newImageH = topImage!.size.height + bottomImage!.size.height
        UIGraphicsBeginImageContextWithOptions(CGSize.init(width: newImageW, height: newImageH), false, 0)
        topImage!.draw(in: CGRect.init(x: 0, y: 0, width: topImage!.size.width, height: topImage!.size.height))
        bottomImage!.draw(in: CGRect.init(x: (newImageW-bottomImage!.size.width)/2.0, y: topImage!.size.height, width: bottomImage!.size.width, height: bottomImage!.size.height))
        let contactImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext();
        return contactImage;
    }
    
}
